crates/worker/src/executor/parameter_server.rs

```diff
@@ -336,3 +303,150 @@ impl JobExecutor for ParameterServerExecutor {
+/// * `device` - The Candle device to perform computations on (e.g., `Device::Cpu`).
+/// * `op` - A closure that takes two tensors and returns a new tensor.
+///
+/// # Returns
+/// A `Result` indicating success or a `TensorOpError` on failure.
+///
+/// # Note
+/// Tensors present in the second file but not the first are ignored. If a tensor
+/// from the first file is not found in the second, it is skipped with a warning.
+/// The `temp_path` will be created and delted by the function. Make sure it doesn't
+/// point to an existing directory that contains important data.
+/// Also make sure that the tensors in `op` are in the same order as they are passed to the function.
+async fn apply_tensor_op<F>(
+    file_a_path: &Path,
+    file_b_path: &Path,
+    output_path: &Path,
+    temp_path: &Path,
+    device: &Device,
+    op: F,
+) -> Result<(), TensorOpError>
+where
+    F: Fn(&Tensor, &Tensor) -> Result<Tensor, candle_core::Error>,
+{
+    // 1. Open both safetensor files in a memory-mapped way.
+    // SAFETY: The MmapedSafetensors::new function is unsafe because it assumes
+    // the underlying file will not be modified while the memory map is active.
+    let tensors_a = unsafe { candle_core::safetensors::MmapedSafetensors::new(file_a_path)? };
+    let tensors_b = unsafe { candle_core::safetensors::MmapedSafetensors::new(file_b_path)? };
+    fs::create_dir_all(temp_path).await?;
+
+    let mut result_tensors = Vec::new();
+
+    // 2. Iterate through each tensor in the first file.
+    for (name, tensor_view) in tensors_a.tensors() {
+        // Try to load the corresponding tensor from the second file.
+        match tensors_b.load(&name, device) {
+            Ok(tensor_b) => {
+                // If found, load the tensor from the first file.
+                let tensor_a = tensor_view.load(device)?;
+
+                // 3. Apply the provided computation function and serialize the result to disk.
+                let result_tensor = op(&tensor_a, &tensor_b)?;
+                let result_path = temp_path.join(name.clone());
+                candle_core::safetensors::save(
+                    &HashMap::from([(name, result_tensor)]),
+                    result_path.clone(),
+                )?;
+                result_tensors.push(result_path);
+            }
+            Err(_) => {
+                // If a tensor from file A doesn't exist in file B, skip it.
+                tracing::warn!("Tensor '{}' not found in second file, skipping.", name);
+                continue;
+            }
+        }
+    }
+
+    // 4. Write all result tensors to the new file.
+    if result_tensors.is_empty() {
+        tracing::warn!("Warning: No matching tensors found to process.");
+    } else {
+        let all_tensors = unsafe { MmapedSafetensors::multi(&result_tensors)? };
+        serialize_to_file(all_tensors.tensors(), &None, output_path)?;
```

Comment:
As discussed offline with @l45k we may want to chunk things into individual tensors and keep it as such while we're aggregating results instead of writing them into one file after each iteration.